### PR TITLE
Optionally show fixes when using `--features ecosystem_ci` with cargo and `--show-fixes` at runtime

### PR DESCRIPTION
**Pull request by konstin**
Example command: `cargo run --bin ruff --features ecosystem_ci -- --ecosystem-ci --no-cache --select F401 --show-fixes crates/ruff/resources/test/fixtures/pyflakes/F401_9.py`

Before, `--show-fixes` was ignored:

```
crates/ruff/resources/test/fixtures/pyflakes/F401_9.py:4:22: F401 [*] `foo.baz` imported but unused
Found 1 error.
[*] 1 potentially fixable with the --fix option.
```

After:

```
crates/ruff/resources/test/fixtures/pyflakes/F401_9.py:4:22: F401 [*] `foo.baz` imported but unused
ℹ Suggested fix
1 1 | """Test: late-binding of `__all__`."""
2 2 | 
3 3 | __all__ = ("bar",)
4   |-from foo import bar, baz
  4 |+from foo import bar

Found 1 error.
[*] 1 potentially fixable with the --fix option.
```

### CODE REVIEW

**Review by MichaReiser — COMMENTED** [review]
I like the addition of showing fixes as part of diagnostics, but I'm not sure if our current implementation is "production-ready," as stated in the source code:

https://github.com/charliermarsh/ruff/blob/b14358fbfe3d59dead9cc3efcded6c0468f2bda3/crates/ruff/src/message/diff.rs#L10-L17

What's your reasoning behind adding a new CLI option vs enabling it when `--show-source` is enabled? My concern is that we start to have too many options and it would be better to just be opinionated about it.

**Comment by MichaReiser** [comment]
> i didn't see the comment you now linked no_mouth. Any preferences where to put it instead? i'd add a (possibly doc(hidden)) --format ecosystem-checks otherwise special-cased similar to the json

Not sure. Could we hide it behind a Rust feature for now? Do we enable all features when building the `debug` artifact on CI?

**Comment by konstin** [comment]
I added `--format ecosystem-ci`, that's the best compromise i could find that also supports `pip install ruff` as baseline

File: crates/ruff/src/settings/types.rs [hunk 1]
```diff
@@ -230,6 +230,10 @@ pub enum SerializationFormat {
     Gitlab,
     Pylint,
     Azure,
+    // We can't make this hidden in clap because clap is an optional dependency
+    // TODO(konstin): Remove ecosystem ci once [ruff::message::diff::Diff] is stable.
+    /// For use with check_ecosystem.py only
+    EcosystemCi,
```
Comment (MichaReiser) [inline crates/ruff/src/settings/types.rs:236]: Is it possible to restrict this variant to a rust feature?

File: crates/ruff_cli/src/printer.rs [hunk 2]
```diff
@@ -178,9 +178,10 @@ impl Printer {
             SerializationFormat::Junit => {
                 JunitEmitter::default().emit(writer, &diagnostics.messages, &context)?;
             }
-            SerializationFormat::Text => {
+            SerializationFormat::Text | SerializationFormat::EcosystemCi => {
                 TextEmitter::default()
                     .with_show_fix_status(show_fix_status(self.autofix_level))
+                    // Currently only has an effect for ecosystem-ci
```
Comment (MichaReiser) [inline crates/ruff_cli/src/printer.rs:184]: What happens if you run `--fix --show-fixes` with the default text emitter? But I guess that would already have been a problem before.

**Review by MichaReiser — APPROVED** [review]
This is okay with me. I think it would be good if we could restrict the variant to a rust feature to not ship in production. 

We could even go as far as to limit the `Diff` implementation to that feature too (or test) to make it clear that the implementation may not be production ready yet.

File: crates/ruff/src/settings/types.rs (hunk 1 above)
Comment (konstin) [reply crates/ruff/src/settings/types.rs:237]: done

File: crates/ruff_cli/src/printer.rs (hunk 2 above)
Comment (konstin) [reply crates/ruff_cli/src/printer.rs]: `--fix --show-fixes` currently has no effect over `--fix`, at least i didn't see any different output

**Comment by konstin** [comment]
Since this broke `cargo test --features all`, i switched to a top level flag that gets patched through by optional fields instead

**Comment by konstin** [comment]
i've restacked this onto #4239 which slightly simplifies the logic

**Comment by konstin** [comment]
> It may be good to update the PR summary and mention that this only applies for builds where the ecosystem feature is enabled.

Thanks for the reminder, updated

**Pull request merged**